src/openrct2/ride/Ride.cpp

```diff
@@ -3881,77 +3881,83 @@ TrackElement* Ride::GetOriginElement(StationIndex stationIndex) const
     {
         log_warning("Invalid ride type for ride %u", id.ToUnderlying());
         return { false };
     }
 
-    if (newStatus != RideStatus::Simulating)
+    window_close_by_number(WindowClass::RideConstruction, id.ToUnderlying());
+
+    auto message = ride_check_for_entrance_exit(id);
+    if (!message.Successful)
     {
-        window_close_by_number(WindowClass::RideConstruction, id.ToUnderlying());
+        ConstructMissingEntranceOrExit();
+        return message;
     }
 
-    auto stationIndexCheck = ride_mode_check_station_present(this);
-    auto stationIndex = stationIndexCheck.StationIndex;
-    if (stationIndex.IsNull())
-        return { false, stationIndexCheck.Message };
+    message = ChangeStatusGetStartElement(trackElement);
+    if (!message.Successful)
+    {
+        return message;
+    }
 
-    auto stationNumbersCheck = ride_mode_check_valid_station_numbers(this);
-    if (!stationNumbersCheck.Successful)
-        return { false, stationNumbersCheck.Message };
+    message = ChangeStatusCheckCompleteCircuit(trackElement);
+    if (!message.Successful)
+    {
+        return message;
+    }
 
-    if (newStatus != RideStatus::Simulating)
+    message = ChangeStatusCheckTrackValidity(trackElement);
+    if (!message.Successful)
     {
-        auto entranceExitCheck = ride_check_for_entrance_exit(id);
-        if (!entranceExitCheck.Successful)
-        {
-            ConstructMissingEntranceOrExit();
-            return { false, entranceExitCheck.Message };
-        }
+        return message;
     }
 
-    // z = ride->stations[i].GetBaseZ();
-    auto startLoc = GetStation(stationIndex).Start;
-    trackElement.x = startLoc.x;
-    trackElement.y = startLoc.y;
-    trackElement.element = reinterpret_cast<TileElement*>(GetOriginElement(stationIndex));
-    if (trackElement.element == nullptr)
+    return ChangeStatusCreateVehicles(isApplying, trackElement);
+}
+
+ResultWithMessage Ride::Simulate(bool isApplying)
+{
+    CoordsXYE trackElement, problematicTrackElement = {};
+    if (type == RIDE_TYPE_NULL)
     {
-        // Maze is strange, station start is 0... investigation required
-        const auto& rtd = GetRideTypeDescriptor();
-        if (!rtd.HasFlag(RIDE_TYPE_FLAG_IS_MAZE))
-            return { false };
+        log_warning("Invalid ride type for ride %u", id.ToUnderlying());
+        return { false };
     }
 
-    if (mode == RideMode::ContinuousCircuit || IsBlockSectioned())
+    auto message = ChangeStatusGetStartElement(trackElement);
+    if (!message.Successful)
     {
-        if (FindTrackGap(trackElement, &problematicTrackElement) && (newStatus != RideStatus::Simulating || IsBlockSectioned()))
-        {
-            ride_scroll_to_track_error(problematicTrackElement);
-            return { false, STR_TRACK_IS_NOT_A_COMPLETE_CIRCUIT };
-        }
+        return message;
+    }
+
+    if (FindTrackGap(trackElement, &problematicTrackElement) && IsBlockSectioned())
```

Comment:
Reverse the order - a block section check is cheaper than finding the track gap.